tools/crm_rule.c

```diff
@@ -0,0 +1,266 @@
+
+    int rc = pcmk_ok;
+    crm_exit_t exit_code = CRM_EX_OK;
+
+    crm_log_cli_init("crm_rule");
+    crm_set_options(NULL, "[options]", long_options,
+                    "Tool for querying the state of rules");
+
+    while (flag >= 0) {
+        flag = crm_get_option(argc, argv, &option_index);
+        switch (flag) {
+            case -1:
+                break;
+
+            case 'V':
+                crm_bump_log_level(argc, argv);
+                break;
+
+            case '$':
+            case '?':
+                crm_help(flag, CRM_EX_OK);
+                break;
+
+            case 'c':
+                rule_mode = crm_rule_mode_check;
+                break;
+
+            case 'd':
+                rule_date = crm_time_new(optarg);
+                if (rule_date == NULL) {
+                    exit_code = CRM_EX_DATAERR;
+                    goto bail;
+                }
+
+                break;
+
+            case 'X':
+                input_xml = optarg;
+                break;
+
+            case 'r':
+                rule_id = strdup(optarg);
+                break;
+
+            default:
+                crm_help(flag, CRM_EX_OK);
+                break;
+        }
+    }
+
+    /* Check command line arguments before opening a connection to
+     * the CIB manager or doing anything else important.
+     */
+    if (rule_mode == crm_rule_mode_check) {
+        if (rule_id == NULL) {
+            CMD_ERR("--check requires use of --rule=\n");
+            crm_help(flag, CRM_EX_USAGE);
+        }
+    }
+
+    /* Set up some defaults. */
+    if (rule_date == NULL) {
+        rule_date = crm_time_new(NULL);
+    }
+
+    /* Where does the XML come from?  If one of various command line options were
+     * given, use those.  Otherwise, connect to the CIB and use that.
+     */
+    if (input_file != NULL) {
+        struct stat sb;
+
+        if (stat(input_file, &sb) == -1) {
+            CMD_ERR("cannot read input file %s: %s", input_file, strerror(errno));
+            exit_code = errno;
+            goto bail;
+        }
+
+        input = filename2xml(input_file);
+    } else if (safe_str_eq(input_xml, "-")) {
+        input = stdin2xml();
+    } else if (input_xml != NULL) {
+        input = string2xml(input_xml);
+    } else {
+        /* Establish a connection to the CIB manager */
+        cib_conn = cib_new();
+        rc = cib_conn->cmds->signon(cib_conn, crm_system_name, cib_command);
+        if (rc != pcmk_ok) {
+            CMD_ERR("Error connecting to the CIB manager: %s", pcmk_strerror(rc));
+            exit_code = rc;
+            goto bail;
+        }
+    }
+
+    /* Now do whichever operation mode was asked for.  There's only one at the
+     * moment so this looks a little silly, but I expect there will be more
+     * modes in the future.
+     */
+    switch(rule_mode) {
+        case crm_rule_mode_check:
+            exit_code = crm_rule_check(cib_conn, input, rule_id, rule_date);
```

Comment:
And this needs a custom mapping, since crm_rule_check() doesn't return either an errno or an exit code

Reply:
Sometimes it does, though - it can return -ENOMEM, and it can return CRM_EX_ERROR.  I guess the return type of that function will need to be dealt with, though I am not sure how just yet.

Reply:
I think I see what I can do.  I'll have a new version in just a minute.